### PR TITLE
Reintroduce domain deprecation

### LINKED ISSUE #1279: Archiving domains
This is functionality that we've had, and taken away, and are re-implementing now:

Add deprecateDomain: https://github.com/JoinColony/colonyNetwork/pull/983
Add support for local skills: https://github.com/JoinColony/colonyNetwork/pull/986
Remove global skills: https://github.com/JoinColony/colonyNetwork/pull/1173

Also this comment: https://github.com/JoinColony/colonyNetwork/pull/986/commits/305266fa82c6cf85da8dec937979b25ae046f868#r740362477

This history has led us into a strange state where each change was path-dependent on the change before. We initially implemented domain deprecation before we had local skills; we tried to update it to use local skills once we had them, which ultimately failed due to unforeseen interactions between colonies, and then went ahead and removed global skills, the feature which required the functionality which made local skill deprecation impossible.

Ultimately, we have sort of come full-circle where we can now implement local skill deprecation as initially envisioned, as there is no longer a need to store _any_ deprecation information on the network itself. However, deploying that change presents some small challenges, namely any _currently-deprecated skills_ will become _undeprecated_, as we will move from network-stored deprecation to colony-stored deprecation, which necessarily has not been set yet.

Speaking to @arrenv, it is unlikely this change will effect many users, if any, as very few users are still active on Gnosis chain, and it is likely that only a small handful of users ever used this functionality in the first place (although this is hard to verify, as the events are recorded on the individual colonies).

### PR DESCRIPTION
**Pull request by kronosapiens**
Closes #1279

This PR restores the ability to deprecate domains and local skills. It also moves skills deprecation entirely to the colony level and away from the network, completing the process begun in #1173.

There are two migration considerations, both of which I believe are relatively minor:

- The current `localSkills` boolean mapping is deprecated and replaced by a `LocalSkill` struct mapping. For already-extant local skills, validity and deprecation behaviors are preserved.
- Due to the "deprecation of deprecation" on the colony network, all currently-deprecated skills will become "undeprecated". Affected users will need to re-deprecate any skills or domains such affected.

The point to emphasize is that these migration issues affect functionality that has been deprecated for some time. As such, the actual effects are likely to be very minor.

TODO:

- [x] Add upgrade test for local skill deprecation

### CODE REVIEW

**Comment by area** [comment]
What on earth has happened to `.storage-layouts-normalized`?

File: contracts/colony/ColonyDataTypes.sol
Comment (area) [inline contracts/colony/ColonyDataTypes.sol:326]: I don't understand why both of these need to be added. Can we not just deprecate a domain by deprecating the associated skill?

Comment (kronosapiens) [reply contracts/colony/ColonyDataTypes.sol:326]: Currently creating a domain does not add an entry to the `LocalSkill`s mapping -- they are parallel mappings.

Comment (area) [reply contracts/colony/ColonyDataTypes.sol:326]: I guess I don't understand what we're doing here then. It sounds like there are two totally distinct pieces of work here: 

* Adding `deprecated` to the domain mapping, which we use to deprecate domains (NB Regardless of implementation, this is the functionality that has to be restored, and what we actually care about).
* Moving tracking of local skills, and whether they are deprecated, to colonies rather than being on the network, with the intention of making things simpler and less gas-intensive.

When you were insistent on point 2 in our original discussions, I assumed it would have an effect on the implementation of point 1, and so together they would be more elegant and consistent. But that doesn't appear to be the case? If this is the sort of implementation you had in mind, why are they bundled together in one PR?

> Currently creating a domain does not add an entry to the LocalSkills mapping

Is there a reason we can't have 'backwards compatability' for such domains, in the same sort of way that you've implemented for e.g. local skill deprecation?

Comment (kronosapiens) [reply contracts/colony/ColonyDataTypes.sol:326]: It's a fair point. In my mind I was looking at "how do we improve our handling of deprecation, given its somewhat random history". Now that it's done, we could conceivably split off the domain part from the rest, but that would keep deprecation fractured. This PR makes deprecation consistent and future proofed for the network. If you'd like me to split the domain piece into a separate PR I can do that, but I think we're best taking it as a bundle.

File: contracts/colonyNetwork/IColonyNetwork.sol
Comment (kronosapiens) [reply contracts/colonyNetwork/IColonyNetwork.sol:103]: Comment updated and moved to here: https://github.com/JoinColony/colonyNetwork/pull/1280/files/22a012bcd8f6dc1c37ac97b5b0fa1390284baf4e#diff-08b59de29cbf4e6c54d624106d18c9a6fd4baaff19ff6a50592df3536095a2d4R101

File: contracts/colony/Colony.sol
Comment (kronosapiens) [reply contracts/colony/Colony.sol:250]: True, but I stylistically liked keeping them separate on the grounds that the latter block is deprecated behavior. Separating them this way more clearly delineates current behavior from what is retained for compatibility.

Comment (area) [reply contracts/colony/Colony.sol:250]: Form should follow function, and this structure is more expensive, gas-wise, for colonies only using the new storage. Characterising the latter functionality as deprecated I would also argue is misleading. This is first-tier functionalty that is migrating a now-deprecated storage layout.

I would like, at the very, very least if there was no way to emit the event twice from calling this function once, which is exactly the sort of thing that could cause the block ingestor grief at some point.

Comment (kronosapiens) [reply contracts/colony/Colony.sol]: Ok, I've made a change. Let me know what you think. I think it's clunkier than what we had before but so it goes.

Comment (area) [reply contracts/colony/Colony.sol]: I had pictured each branch in the if statement if having its own `emit`, if that makes things less clunky for you?

File: contracts/colony/ColonyDataTypes.sol
Comment (area) [inline contracts/colony/ColonyDataTypes.sol:321]: Having trouble writing a test to demonstrate it right now (IMWSS doesn't play ball with the 'historical' deployment scripts right now, it seems), but while this change is fine standalone, in the context of an already deployed system, breaks anything that calls `getDomain` on a colony. That's a lot of extensions, but also the `ReputationMiningCycle`, so this can't be merged until that's solved (and, ideally, a test written).

Comment (area) [reply contracts/colony/ColonyDataTypes.sol:321]: After some experimenting, I think going this way is okay, and it's only removing elements from such a struct that would cause a revert to occur? I used this to tinker:

```
pragma solidity >=0.7.0 <0.9.0;

contract Test1 {
  struct Domain {
    uint256 skillId;
    uint256 fundingPotId;
  }

  function x() public pure returns (Domain memory domain) {
    return Domain(1, 1);
  }
}

contract Test2 {
  struct Domain {
    uint256 skillId;
    uint256 fundingPotId;
    bool deprecated;
  }

    function x() public pure returns (Domain memory domain) {
        return Domain(2, 2, false);
    }
}

contract Test3 {
    function test(address a) public pure returns (uint256) {
        Test1.Domain memory d = Test1(a).x();
        return d.skillId;
    }
}

contract Test4 {
    function test(address a) public pure returns (uint256) {
        Test2.Domain memory d = Test2(a).x();
        return d.skillId;
    }
}
```

I would still very much like a test using our contracts demonstrating that's the case, though - bricking the mining cycle, in particular, is unacceptable.

**Review by area — COMMENTED** [review]
Implementation looks good. Just need to work out how to convince myself further that it won't brick everything with the return data change when calling `getDomain`.

File: test/contracts-network/colony-expenditure.js
Comment (kronosapiens) [reply test/contracts-network/colony-expenditure.js:368]: Fair point, at this point in the test the colony and network are both current.

**Comment by area** [comment]
I spent this afternoon getting extremely comfortable with Remix and EVM Opcodes to convince myself further that returning too much data in a call is okay, and I believe it is. I might end up adding to the relevant OneTxPayment test though.

I also don't believe bricking the mining cycle is a risk after all, as the only `getDomain()` call is in the initialization, so that makes me a little happier. I still don't want to merge this until after multisig has been merged and deployed, though.

**Comment by area** [comment]
Still waiting on multisig.

**Comment by area** [comment]
> Still waiting on multisig.

That's now merged, but not yet deployed... so closer, at least!

**Comment by area** [comment]
Needs #1299 merged, and then versions bumped accordingly.

**Review by area — APPROVED** [review]

**Pull request merged**
